lib/composite_cache_store.rb

```diff
@@ -4,146 +4,157 @@
+    layers.each { |layer| layer.delete_matched(...) }
   end
 
   def delete_multi(...)
-    layers.each { |store| store.delete_multi(...) }
+    layers.map { |layer| layer.delete_multi(...) }.last
   end
 
   def exist?(...)
-    layers.each do |store|
-      return true if store.exist?(...)
-    end
-    false
+    layers.any? { |layer| layer.exist?(...) }
   end
 
-  def fetch(*args, &block)
-    f = ->(store) do
-      return store.fetch(*args, &block) if store == layers.last
-      store.fetch(*args) { f.call(layers[layers.index(store) + 1]) }
+  def fetch(name, options = nil, &block)
+    options ||= {}
+
+    if options[:force]
+      raise ArgumentError, "Missing block: Calling `Cache#fetch` with `force: true` requires a block." unless block
+      value = block&.call(name)
+      layers.each { |layer| layer.write(name, options) }
+      return value
     end
-    f.call(layers.first)
-  end
 
-  def fetch_multi(*args, &block)
-    fm = ->(store) do
-      return store.fetch_multi(*args, &block) if store == layers.last
-      store.fetch_multi(*args) { fm.call(layers[layers.index(store) + 1]) }
+    read(name, options) do |value, warm_layer, cold_layers|
+      value ||= block&.call(name) unless warm_layer
+      cold_layers.each do |layer|
+        layer.write(name, value, options) unless value.nil? && options[:skip_nil]
+      end
+      return value
     end
-    fm.call(layers.first)
   end
 
-  def increment(...)
-    layers.each { |store| store.increment(...) }
+  def fetch_multi(*names, &block)
+    raise ArgumentError, "Missing block: `Cache#fetch_multi` requires a block." unless block
+
+    keys = names.dup
+    options = keys.extract_options!
+
+    if options[:force]
+      value = keys.each_with_object({}) { |key, memo| memo[key] = block&.call(key) }
+      layers.each { |layer| layer.write_multi(value, options) }
+      return value
+    end
+
+    read_multi(*names) do |value, warm_layer, cold_layers|
+      missing_keys = keys - value.keys
+      missing_keys.each { |key| value[key] = block&.call(key) }
+
+      cold_layers.each do |layer|
+        value.compact! if options[:skip_nil]
+        layer.write_multi(value, options)
+      end
+
+      return keys.each_with_object({}) { |key, memo| memo[key] = value[key] }
+    end
   end
 
   def mute
-    m = ->(store) do
-      return store.mute { yield } if store == layers.last
-      store.mute { m.call(layers[layers.index(store) + 1]) }
-    end
-    m.call(layers.first)
+    layers.each { |layer| layer.mute { yield } }
   end
 
-  def read(*args)
-    r = ->(store) do
-      return store.read(*args) if store == layers.last
-      store.fetch(*args) { r.call(layers[layers.index(store) + 1]) }
+  def read(name, options = nil)
+    warm_layer = nil
+    cold_layers = []
+    value = nil
+
+    layers.each do |layer|
+      value = layer.read(name, options)
+      warm_layer = layer if !value.nil? || layer.exist?(name, options)
```

Comment:
I think you're testing `layer.exist?(name)` here just in case `nil` was the value returned, because we can't tell whether `nil` indicates a cache miss or an actual `nil`? To do that, you're hitting every intermediate cache twice before you get to the warm layer. That feels like a significant performance hit, right?

Reply:
Definitely a hot spot. I've updated `read` to use [`ActiveSupport::Cache::Strategy::LocalCache#with_local_cache`](https://github.com/rails/rails/blob/5cc5556ef3e55a7a8efba8290dff7ba06d227e99/activesupport/lib/active_support/cache/strategy/local_cache.rb#L62-L65) to mitigate this problem.